presto-pinot-toolkit/src/main/java/com/facebook/presto/pinot/query/PinotQueryGeneratorContext.java

```diff
@@ -150,8 +154,18 @@ public PinotQueryGeneratorContext withAggregation(
             int aggregations,
             Set<VariableReferenceExpression> hiddenColumnSet)
     {
-        // there is only one aggregation supported.
-        checkSupported(!hasAggregation(), "Pinot doesn't support aggregation on top of the aggregated data");
+        // there is only one aggregation supported unless distinct is used.
+        boolean passed = false;
+        Collection<Selection> selections = newSelections.values();
+        for (Selection selection : selections) {
+            if (selection.getDefinition().startsWith(DISTINCT_COUNT_FUNCTION_NAME.toUpperCase(Locale.ENGLISH))) {
+                hiddenColumnSet = Collections.emptySet();
+                passed = true;
+            }
+        }
+        if (!passed) {
+            checkSupported(!hasAggregation(), "Pinot doesn't support aggregation on top of the aggregated data");
+        }
```

Comment:
The hiddenColumnSet should be changed here, such that the loop above becomes read-only. 

Also, would this check make sense to guard the overwriting of hiddenColumnSet:

`if (!passed) {
   checkSupported(!hashAggregation(), "....");
 } else if (hasAggregation()) {
   hiddenColumnSet = ImmutableSet.of();
 }
`

Also, please use either ImmutableSet.of() if the intention is for hiddenColumnSet to be never added to again for the life of this context. Or use a new HashSet<> if you want to allow future additions. Presto does not encourage Collections.emptySet. 

I wonder if it is possible to have another aggregation on top of a distinct ? Like can we have three aggregations stacked together ? What should we do in that case ? The safest thing would be to detect and back out.

Reply:
Agreed.
If it's triple aggregation, then the most outer layer will check and found out there is an aggregation on top of an aggregation, then fail.

Reply:
Refined this logic.

Reply:
Thanks !. Reads better now :-)